src/main/java/com/teragrep/pth10/ast/DefaultTimeFormat.java

```diff
@@ -112,6 +131,13 @@ else if (attempt == 1) {
     private Date parseDate(String time, String timeFormat) throws ParseException {
         SimpleDateFormat sdf = new SimpleDateFormat(timeFormat);
         sdf.setLenient(false);
-        return sdf.parse(time);
+        Date date = sdf.parse(time);
+
+        // If date is for latest timeQualifier and has fractions-of-second, add 1 second to capture events
+        // that are on the same second
+        if (date.toInstant().getNano() > 0 && isLatest) {
+            date = Date.from(date.toInstant().plus(1, ChronoUnit.SECONDS));
+        }
+        return date;
     }
```

Comment:
I had couple of ideas for this logic and I came up with this. I think this could just be replaced by having a list of formats and a for loop that tries to parse each format and returns the date if no exception was thrown. if the end of the for loop is reached an exception is thrown. this would also leave room for easily adding or removing formats in the future. List of formats could be given in the constructor and a secondary constructor could have a default list of formats. Here is an example of code

```
public Date parse(final String time) {
        // Try parsing all formats
        for (String format: formats) {
            try {
                return parseDate(time, format);
            } catch (ParseException ignored) {
            }
        }
        throw new RuntimeException("Could not parse time <" + time + ">");
    } 
```
you could keep track of the parser error and return the latest one if you think it is important information

Reply:
fixed